lib/view_component/base.rb

```diff
@@ -246,11 +248,37 @@ def with_collection_parameter(param)
       end
 
       def collection_parameter_name
-        (@with_collection_parameter || name.demodulize.underscore.chomp("_component")).to_sym
+        (
+          with_collection_parameter_attr ||
+          name.demodulize.underscore.chomp("_component")
+        ).to_sym
+      end
+
+      # Ensure the component initializer accepts the
+      # collection parameter. By default, we do not
+      # validate that the default parameter name
+      # is accepted, as support for collection
+      # rendering is optional.
+      def validate_collection_parameter!(validate_default: false)
+        parameter = validate_default ? collection_parameter_name : with_collection_parameter_attr
+
+        return unless parameter
+        return if instance_method(:initialize).parameters.map(&:last).include?(parameter)
+
+        raise ArgumentError.new(
+          "#{self} initializer must accept " \
+          "`#{parameter}` collection parameter."
+        )
       end
 
       private
 
+      # We use #with_collection_parameter as a setter,
+      # so this method is suffixed with _attr.
+      def with_collection_parameter_attr
+        @with_collection_parameter
+      end
```

Comment:
would it be better to just have the instance variable be `@collection_parameter` ?